[Backtracing][Linux] Enable Linux backtracing, add tests.

al45tair opened the pull request:
Turn on the Linux backtracing implementation and update the tests.

rdar://101623265

al45tair committed:
[Backtracing][IRGen] Add a semantic attribute to force frame pointer.

The Swift backtracer's frame pointer unwinder cannot work on Linux
without this change, because the compiler omits the frame pointer from
the function in libSwift_Backtracing that actually captures the stack.

rdar://110260855

al45tair committed:
[Backtracing][Linux] Reorganise modules for _Backtracing.

Using `SwiftShims` is undesirable - it creates all kinds of build issues,
and means shipping the `_SwiftBacktracing.h` header in the SDK, which is
not necessary.

While we're doing this, add the necessary definitions for reading ELF
and DWARF information.

rdar://110261712

al45tair committed:
[Backtracing][Linux] Add Linux support to the _Backtracing module.

Use the new module structure rather the old SwiftShims header.  This
is much cleaner and lets us include operating system headers to get
the relevant definitions where possible.

Add code to support ELF and DWARF, including decompression using
zlib, zstd and liblzma if those turn out to be required and available.

rdar://110261712

al45tair committed:
[Backtracing][Linux] Add Linux support to swift-backtrace.

We need a Linux specific `Target` implementation, and a couple of minor
tweaks to make things build everywhere.

rdar://110262673

al45tair committed:
[Backtracing][Linux] Enable Linux backtracing, add tests.

Turn on the Linux backtracing implementation and update the tests.

rdar://101623265

al45tair commented:
This needs #66337, which needs #66335 (which in turn requires #66334 and #66333).

al45tair commented:
@swift-ci Please test

al45tair committed:
[Backtracing][Linux] Temporarily add a default addressWidth.

This is for compatibility, so that I can split up the PRs.
We'll remove it in the next PR.

rdar://110261712

al45tair committed:
Revert "[Backtracing][Linux] Temporarily add a default addressWidth."

This reverts commit 0d7c45d9dac68ca820baec816b8857070168136a.

al45tair committed:
[Backtracing][Linux] Enable backtrace-on-crash for Linux.

This reverts commit e5f4c2015e464db9b3af7da1e969102dbafd6076.

rdar://101623265

al45tair commented:
@swift-ci Please test

al45tair committed:
[Backtracing][Linux][Tests] Fix verify_all_overlays test.

The `_Backtracing` module has a number of private implementation only
imports that aren't used outside of the module and that don't require
any additional libraries (hence they aren't relevant to the outside
world).  `verify_all_overlays.py` needs to know about these when it
does its test, because it loadas the module as the main module, which
results in implementation only imports being required instead of
ignored.

rdar://110261712

al45tair commented:
@swift-ci Please test

al45tair commented:
When reviewing, you can start at c9992ad; previous change sets are from #66333, #66334, #66335 and #66337 and will be reviewed there.

MaxDesiatov commented on include/swift/Runtime/Backtrace.h:
```diff
@@ -123,8 +134,15 @@ struct BacktraceSettings {
+SWIFT_RUNTIME_STDLIB_SPI
+bool _swift_backtrace_isThunkFunction(const char *mangledName);
 
+SWIFT_RUNTIME_STDLIB_SPI
```
Would this benefit from a doc comment that clarifies the use of each parameter and the return pointer as well?

al45tair replied:
There's a fairly extensive doc comment for it in the `.cpp` file already; I'd rather not duplicate it here, and as this is SPI anyway I'm not especially fussed about it being here versus there.

al45tair replied:
(This is also not really the right PR to review this particular code :-))

MaxDesiatov commented on include/swift/Runtime/CrashInfo.h:
```diff
@@ -18,22 +18,52 @@
 extern "C" {
 #endif
 
 struct CrashInfo {
```
Ditto, a doc comment would help to clarify what `mctx` is, while for other fields additional context would also be helpful.

MaxDesiatov commented on include/swift/Runtime/CrashInfo.h:
```diff
@@ -18,22 +18,52 @@
+  /* Then len bytes of data */
+};
+
+struct thread {
```
This would benefit from a doc comment that clarifies what `next` and `uctx` are for. I'd assume `tid` refers to "thread id", but a doc comment would confirm that too.

al45tair replied:
As is this, I think (#66334).

MaxDesiatov commented on stdlib/public/Backtracing/ArrayImageSource.swift:
```diff
@@ -0,0 +1,50 @@
+
+import Swift
+
+@_implementationOnly import OS.Libc
```
I wasn't aware of this module, do we finally have a replacement for all the numerous platform-specific libc modules?

al45tair replied:
That is relevant here because of 1c8362b, which was needed because I did that (and because the test does slightly the wrong thing — it treats the module it's testing as the main module, which causes the deserialiser to behave differently).

MaxDesiatov commented on stdlib/public/Backtracing/Compression.swift:
```diff
@@ -0,0 +1,528 @@
+  init() {}
+
+  func decompress(input: InputSource, output: OutputSink) throws -> UInt {
+
```
formatting nit: stray newline

al45tair replied:
And that's for #66335.

al45tair commented:
Just to clarify, the commits that need reviewing in this PR are c9992ad, e3425aa and 1c8362b. The others aren't relevant to this one, and I don't want to get everything all mixed up here if we can help it.

al45tair commented:
https://github.com/apple/swift-corelibs-xctest/pull/445

@swift-ci Please test Linux platform

al45tair commented:
https://github.com/apple/swift-corelibs-xctest/pull/445

@swift-ci Please test Linux platform

al45tair commented:
https://github.com/apple/swift-corelibs-xctest/pull/445

@swift-ci Please clean test Linux platform

al45tair committed:
[Backtracing][Linux] Work correctly in the presence of partial reads.

There's a chance that pipes might perform a partial read; we should
handle that case.

rdar://110261712

al45tair committed:
[Backtracing][Linux] Remove some stray "public" annotations.

We don't need these here.  They were a leftover from previous code.

rdar://110262673

al45tair commented:
https://github.com/apple/swift-corelibs-xctest/pull/445

@swift-ci Please clean test Linux platform

al45tair commented:
https://github.com/apple/swift-corelibs-xctest/pull/445

@swift-ci Please clean test Linux platform

al45tair committed:
[Backtracing][Linux] Remove status argument.

The `status` argument to the `_swift_backtrace_demangle()` function
isn't especially useful, won't match behaviour on Windows, and we
actually don't use it in the Swift code that calls this SPI.

Remove it.

rdar://110261712

al45tair committed:
[Backtracing][Linux] Add a comment about __cxa_demangle.

`__cxa_demangle()` is a rather unusual API; one of its "features" is that
the pointer you pass in must either be `nullptr`, in which case it
will call `malloc()` itself, _or_ it has to be a pointer to a block
of memory allocated with `malloc()`, because `__cxa_demangle()` may
`realloc()` it for you.

This seems to me to be something of a non-obvious footgun, so we never
pass the caller's pointer through to `__cxa_demangle()`, which lets them
decide how they want to allocate space.

rdar://110261712

al45tair commented:
Ah, the `swift-corelibs-xctest` PR is merged now, which will upset that test run.

al45tair commented:
@swift-ci Please test

al45tair committed:
[Backtracing][Linux] Also remove the status argument from Runtime module.

There's a separate declaration here because we can't include the `Backtrace.h`
header from inside `modules/Runtime/Runtime.h`.

rdar://110261712

al45tair commented:
@swift-ci Please test
